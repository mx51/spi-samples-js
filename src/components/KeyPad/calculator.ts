export default (amount: string, num: string): string => {
  switch (num) {
    case 'Backspace':
      if (amount.length === 1) return '0';
      return amount.substring(0, amount.length - 1);
    case '.':
      return amount.includes('.') ? amount : amount + num;
    default:
      if (amount === '0') return num;
      if (amount.includes('.') && amount.indexOf('.') === amount.length - 3) return amount;
      if (!amount.includes('.') && amount.length === 4) return amount;
      return amount + num;
  }
};
