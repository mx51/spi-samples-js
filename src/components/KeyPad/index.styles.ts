import { makeStyles, Theme, createStyles } from '@material-ui/core/styles';

export default makeStyles((theme: Theme) =>
  createStyles({
    root: {
      height: `calc(100vh - 48px)`,
      position: 'sticky',
      top: theme.spacing(6),
    },
    heading: {
      padding: theme.spacing(2),
    },
    grid: {
      flexGrow: 1,
    },
    keypadDisplay: {
      padding: theme.spacing(2),
    },
    keypadDisplayLabel: {
      textAlign: 'center',
      fontSize: '1rem',
      color: theme.palette.text.secondary,
    },
    keypadDisplayAmount: {
      textAlign: 'right',
      fontSize: '4.5rem',
      fontWeight: 'normal',
    },
    keypadBtn: {
      height: '100%',
      borderRadius: 0,
      fontSize: '2rem',
      borderTopColor: 'transparent',
      borderLeftColor: 'transparent',
      borderRightColor: theme.palette.divider,
      borderBottomColor: theme.palette.divider,
    },
    keypadBtnLabel: {
      fontSize: '3rem',
      fontWeight: 400,
      color: theme.palette.text.secondary,
    },
    okBtn: {
      padding: theme.spacing(2),
      borderRadius: 0,
      backgroundColor: theme.palette.error.dark,
      '&:hover, &:active, &:focus': {
        background: theme.palette.error.main,
        color: theme.palette.common.white,
      },
    },
    okBtnLabel: {
      fontWeight: 500,
      fontSize: '1.5rem',
      lineHeight: '1.5rem',
    },
  })
);
