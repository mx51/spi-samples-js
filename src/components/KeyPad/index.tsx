import KeyPad from './KeyPad';

export default KeyPad;
