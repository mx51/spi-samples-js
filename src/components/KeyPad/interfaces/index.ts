export interface IKeyPadProps {
  title: string;
  subtitle: string;
  defaultAmount: number;
  onAmountChange: (amount: number) => void;
  open: boolean;
  onClose: () => void;
}
