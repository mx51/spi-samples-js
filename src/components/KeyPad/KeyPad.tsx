import React, { useState, useEffect } from 'react';
import { Box, Button, Divider, Grid, Paper, Typography } from '@material-ui/core';
import BackspaceOutlinedIcon from '@material-ui/icons/BackspaceOutlined';
import CloseIcon from '@material-ui/icons/Close';
import calculator from './calculator';
import useStyles from './index.styles';
import { IKeyPadProps } from './interfaces';

function KeyPad({ title, subtitle, defaultAmount, onAmountChange, onClose }: IKeyPadProps): React.ReactElement {
  const classes = useStyles();
  const KEYS = ['1', '2', '3', '4', '5', '6', '7', '8', '9', 'Backspace', '0', '.'];
  const [amount, setAmount] = useState((defaultAmount / 100).toString());

  const amountChangeHandler = () => {
    const newAmount = Math.round(parseFloat(amount) * 100);
    if (newAmount !== defaultAmount) {
      onAmountChange(newAmount);
    } else {
      onClose();
    }
  };

  const updateAmount = (num: string) => {
    if (KEYS.includes(num)) {
      setAmount((amt) => calculator(amt, num));
    } else if (num === 'Enter') {
      amountChangeHandler();
    }
  };

  const keyboardHandler = (event: KeyboardEvent) => {
    updateAmount(event.key);
  };

  useEffect(() => {
    document.addEventListener('keydown', keyboardHandler);

    return function cleanup() {
      document.removeEventListener('keydown', keyboardHandler);
    };
  }, [amount]);

  return (
    <Box component={Paper} className={classes.root} display="flex" flexDirection="column">
      <Box display="flex" justifyContent="space-between" className={classes.heading} alignItems="center">
        <Typography variant="h6" component="h1">
          {title}
        </Typography>
        <Box>
          <Button aria-label="Close Button" onClick={onClose}>
            <CloseIcon />
          </Button>
        </Box>
      </Box>
      <Divider />
      <Box className={classes.keypadDisplay}>
        <Typography className={classes.keypadDisplayLabel}>{subtitle}</Typography>
        <Typography variant="h6" component="h1" className={classes.keypadDisplayAmount}>
          {amount}
        </Typography>
      </Box>
      <Divider />
      <Grid container className={classes.grid}>
        {KEYS.map((num) => (
          <Grid item xs={4} key={`keypad-${num}`}>
            <Button
              color="primary"
              variant="outlined"
              fullWidth
              classes={{ root: classes.keypadBtn, label: classes.keypadBtnLabel }}
              onClick={() => updateAmount(num)}
            >
              {num !== 'Backspace' ? num : <BackspaceOutlinedIcon className={classes.keypadBtnLabel} />}
            </Button>
          </Grid>
        ))}
      </Grid>
      <Divider />
      <Button
        variant="contained"
        color="primary"
        size="large"
        classes={{ root: classes.okBtn, label: classes.okBtnLabel }}
        onClick={amountChangeHandler}
      >
        OK
      </Button>
    </Box>
  );
}

export default KeyPad;
