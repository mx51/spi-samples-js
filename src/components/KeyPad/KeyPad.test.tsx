import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import '@testing-library/jest-dom/extend-expect';
import KeyPad from '.';

describe('Test <KeyPad />', () => {
  test('snapshot for KeyPad component', () => {
    const onAmountChange = jest.fn();
    const onClose = jest.fn();

    const { container } = render(
      <KeyPad
        title="Surcharge"
        subtitle="Enter surcharge amount"
        defaultAmount={0}
        onAmountChange={onAmountChange}
        open={false}
        onClose={onClose}
      />
    );
    expect(container).toMatchSnapshot();
  });

  test('Ok button clicked without amount change', () => {
    const onAmountChange = jest.fn();
    const onClose = jest.fn();
    const { getByText } = render(
      <KeyPad
        title="Surcharge"
        subtitle="Enter surcharge amount"
        defaultAmount={50}
        onAmountChange={onAmountChange}
        open={false}
        onClose={onClose}
      />
    );

    fireEvent.click(getByText(/ok/i));

    expect(onAmountChange.mock.calls.length).toBe(0);
    expect(onClose.mock.calls.length).toBe(1);
  });

  test('Amount change on keypad & Ok button click', () => {
    const onAmountChange = jest.fn();
    const onClose = jest.fn();
    const { getByText } = render(
      <KeyPad
        title="Surcharge"
        subtitle="Enter surcharge amount"
        defaultAmount={0}
        onAmountChange={onAmountChange}
        open={false}
        onClose={onClose}
      />
    );

    fireEvent.click(getByText(/7/i));
    fireEvent.click(getByText(/2/i));
    fireEvent.click(getByText(/ok/i));

    expect(onAmountChange.mock.calls.length).toBe(1);
    expect(onAmountChange.mock.calls[0][0]).toBe(7200);
  });

  test('Amount change on keydown & Enter', () => {
    const onAmountChange = jest.fn();
    const onClose = jest.fn();
    const { container } = render(
      <KeyPad
        title="Surcharge"
        subtitle="Enter surcharge amount"
        defaultAmount={0}
        onAmountChange={onAmountChange}
        open={false}
        onClose={onClose}
      />
    );

    fireEvent.keyDown(container, { key: '4' });
    fireEvent.keyDown(container, { key: '.' });
    fireEvent.keyDown(container, { key: '5' });
    fireEvent.keyDown(container, { key: 'Enter' });

    expect(onAmountChange.mock.calls.length).toBe(1);
    expect(onAmountChange.mock.calls[0][0]).toBe(450);
  });
});
