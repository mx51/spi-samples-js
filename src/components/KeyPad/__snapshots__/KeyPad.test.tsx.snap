// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <KeyPad /> snapshot for KeyPad component 1`] = `
<div>
  <div
    class="MuiPaper-root MuiBox-root MuiBox-root-11 makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiBox-root MuiBox-root-12 makeStyles-heading-2"
    >
      <h1
        class="MuiTypography-root MuiTypography-h6"
      >
        Surcharge
      </h1>
      <div
        class="MuiBox-root MuiBox-root-13"
      >
        <button
          aria-label="Close Button"
          class="MuiButtonBase-root MuiButton-root MuiButton-text"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <hr
      class="MuiDivider-root"
    />
    <div
      class="MuiBox-root MuiBox-root-14 makeStyles-keypadDisplay-4"
    >
      <p
        class="MuiTypography-root makeStyles-keypadDisplayLabel-5 MuiTypography-body1"
      >
        Enter surcharge amount
      </p>
      <h1
        class="MuiTypography-root makeStyles-keypadDisplayAmount-6 MuiTypography-h6"
      >
        0
      </h1>
    </div>
    <hr
      class="MuiDivider-root"
    />
    <div
      class="MuiGrid-root makeStyles-grid-3 MuiGrid-container"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-4"
      >
        <button
          class="MuiButtonBase-root MuiButton-root makeStyles-keypadBtn-7 MuiButton-outlined MuiButton-outlinedPrimary MuiButton-fullWidth"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label makeStyles-keypadBtnLabel-8"
          >
            1
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-4"
      >
        <button
          class="MuiButtonBase-root MuiButton-root makeStyles-keypadBtn-7 MuiButton-outlined MuiButton-outlinedPrimary MuiButton-fullWidth"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label makeStyles-keypadBtnLabel-8"
          >
            2
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-4"
      >
        <button
          class="MuiButtonBase-root MuiButton-root makeStyles-keypadBtn-7 MuiButton-outlined MuiButton-outlinedPrimary MuiButton-fullWidth"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label makeStyles-keypadBtnLabel-8"
          >
            3
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-4"
      >
        <button
          class="MuiButtonBase-root MuiButton-root makeStyles-keypadBtn-7 MuiButton-outlined MuiButton-outlinedPrimary MuiButton-fullWidth"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label makeStyles-keypadBtnLabel-8"
          >
            4
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-4"
      >
        <button
          class="MuiButtonBase-root MuiButton-root makeStyles-keypadBtn-7 MuiButton-outlined MuiButton-outlinedPrimary MuiButton-fullWidth"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label makeStyles-keypadBtnLabel-8"
          >
            5
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-4"
      >
        <button
          class="MuiButtonBase-root MuiButton-root makeStyles-keypadBtn-7 MuiButton-outlined MuiButton-outlinedPrimary MuiButton-fullWidth"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label makeStyles-keypadBtnLabel-8"
          >
            6
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-4"
      >
        <button
          class="MuiButtonBase-root MuiButton-root makeStyles-keypadBtn-7 MuiButton-outlined MuiButton-outlinedPrimary MuiButton-fullWidth"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label makeStyles-keypadBtnLabel-8"
          >
            7
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-4"
      >
        <button
          class="MuiButtonBase-root MuiButton-root makeStyles-keypadBtn-7 MuiButton-outlined MuiButton-outlinedPrimary MuiButton-fullWidth"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label makeStyles-keypadBtnLabel-8"
          >
            8
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-4"
      >
        <button
          class="MuiButtonBase-root MuiButton-root makeStyles-keypadBtn-7 MuiButton-outlined MuiButton-outlinedPrimary MuiButton-fullWidth"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label makeStyles-keypadBtnLabel-8"
          >
            9
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-4"
      >
        <button
          class="MuiButtonBase-root MuiButton-root makeStyles-keypadBtn-7 MuiButton-outlined MuiButton-outlinedPrimary MuiButton-fullWidth"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label makeStyles-keypadBtnLabel-8"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-keypadBtnLabel-8"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M22 3H7c-.69 0-1.23.35-1.59.88L0 12l5.41 8.11c.36.53.9.89 1.59.89h15c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16H7.07L2.4 12l4.66-7H22v14zm-11.59-2L14 13.41 17.59 17 19 15.59 15.41 12 19 8.41 17.59 7 14 10.59 10.41 7 9 8.41 12.59 12 9 15.59z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-4"
      >
        <button
          class="MuiButtonBase-root MuiButton-root makeStyles-keypadBtn-7 MuiButton-outlined MuiButton-outlinedPrimary MuiButton-fullWidth"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label makeStyles-keypadBtnLabel-8"
          >
            0
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-4"
      >
        <button
          class="MuiButtonBase-root MuiButton-root makeStyles-keypadBtn-7 MuiButton-outlined MuiButton-outlinedPrimary MuiButton-fullWidth"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label makeStyles-keypadBtnLabel-8"
          >
            .
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <hr
      class="MuiDivider-root"
    />
    <button
      class="MuiButtonBase-root MuiButton-root makeStyles-okBtn-9 MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label makeStyles-okBtnLabel-10"
      >
        OK
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
